action/protocol/staking/handler_stake_migrate.go

```diff
@@ -0,0 +1,181 @@
+	actLogs := make([]*action.Log, 0)
+	transferLogs := make([]*action.TransactionLog, 0)
+	si := csm.SM().Snapshot()
+	revertSM := func() {
+		if revertErr := csm.SM().Revert(si); revertErr != nil {
+			log.L().Panic("failed to revert state", zap.Error(revertErr))
+		}
+	}
+
+	// validate bucket index
+	bucket, rErr := p.fetchBucket(csm, act.BucketIndex())
+	if rErr != nil {
+		return nil, nil, 0, rErr
+	}
+	if err := p.validateStakeMigrate(ctx, bucket, csm); err != nil {
+		return nil, nil, 0, err
+	}
+
+	// force-withdraw native bucket
+	staker, rerr := fetchCaller(ctx, csm, big.NewInt(0))
+	if rerr != nil {
+		return nil, nil, 0, errors.Wrap(rerr, "failed to fetch caller")
+	}
+	candidate := csm.GetByIdentifier(bucket.Candidate)
+	if candidate == nil {
+		return nil, nil, 0, errCandNotExist
+	}
+	actLog, tLog, err := p.withdrawBucket(ctx, staker, bucket, candidate, csm)
+	if err != nil {
+		return nil, nil, 0, err
+	}
+	actLogs = append(actLogs, actLog.Build(ctx, nil))
+	transferLogs = append(transferLogs, tLog)
+
+	// call staking contract to stake
+	duration := int64(bucket.StakedDuration / p.getBlockInterval(protocol.MustGetBlockCtx(ctx).BlockHeight))
+	excReceipt, err := p.createNFTBucket(ctx, act, bucket.StakedAmount, big.NewInt(duration), candidate, csm.SM())
+	if err != nil {
+		revertSM()
+		return nil, nil, 0, errors.Wrap(err, "failed to handle execution action")
+	}
+	if excReceipt.Status != uint64(iotextypes.ReceiptStatus_Success) {
+		// TODO: return err or handle error?
+		revertSM()
+		return nil, nil, 0, errors.Errorf("execution failed with status %d", excReceipt.Status)
+	}
+	// add sub-receipts logs
+	actLogs = append(actLogs, excReceipt.Logs()...)
+	transferLogs = append(transferLogs, excReceipt.TransactionLogs()...)
+	return actLogs, transferLogs, excReceipt.GasConsumed, nil
+}
+
+func (p *Protocol) validateStakeMigrate(ctx context.Context, bucket *VoteBucket, csm CandidateStateManager) error {
+	if err := validateBucketOwner(bucket, protocol.MustGetActionCtx(ctx).Caller); err != nil {
+		return err
+	}
+	if !bucket.AutoStake {
+		return &handleError{
+			err:           errors.New("cannot migrate non-auto-staked bucket"),
+			failureStatus: iotextypes.ReceiptStatus_ErrInvalidBucketType,
+		}
+	}
+	if bucket.isUnstaked() {
+		return &handleError{
+			err:           errors.New("cannot migrate unstaked bucket"),
+			failureStatus: iotextypes.ReceiptStatus_ErrInvalidBucketType,
+		}
+	}
+	if err := validateBucketWithoutEndorsement(NewEndorsementStateManager(csm.SM()), bucket, protocol.MustGetBlockCtx(ctx).BlockHeight); err != nil {
+		return err
+	}
+	return validateBucketSelfStake(protocol.MustGetFeatureCtx(ctx), csm, bucket, false)
+}
+
+func (p *Protocol) withdrawBucket(ctx context.Context, withdrawer *state.Account, bucket *VoteBucket, cand *Candidate, csm CandidateStateManager) (*receiptLog, *action.TransactionLog, error) {
```

Comment:
share with withdraw bucket action

Reply:
This may not apply to handle withdraw action, as it involves the reduction of votes which is handled in unstake action